src/Common/BackgroundShellCommandHolder.cpp

```diff
@@ -0,0 +1,37 @@
+#include <Common/logger_useful.h>
+#include <Common/Exception.h>
+#include <Common/BackgroundShellCommandHolder.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+LoggerPtr BackgroundShellCommandHolder::getLogger()
+{
+    return ::getLogger("BackgroundShellCommandHolder");
+}
+
+
+void BackgroundShellCommandHolder::removeCommand(pid_t pid)
+{
+    std::lock_guard lock(mutex);
+    bool is_erased = active_shell_commands.erase(pid);
+    LOG_TRACE(getLogger(), "Try to erase command with the pid {}, is_erased: {}", pid, is_erased);
+}
+
+void BackgroundShellCommandHolder::addCommand(std::unique_ptr<ShellCommand> command)
+{
+    std::lock_guard lock(mutex);
+    pid_t command_pid = command->getPid();
+
+    auto [iterator, is_inserted] = active_shell_commands.emplace(std::make_pair(command_pid, std::move(command)));
```

Comment:
If item with the pid is already in the map, new command will be destroyed here (after moved from). (To not happen this use `try_emplace`) 

It looks like it can lead to hanging in `tryWait` in dtr of `ShellCommand` here if It will not be killed depends on `terminate_in_destructor_strategy`.

It makes sense to handle this case more carefully.
Also it is possible that old command is present but already terminated and waited.

Reply:
>If item with the pid is already in the map, new command will be destroyed here (after moved from). (To not happen this use try_emplace)

Yes, actually I haven't any better solution right now, for case when we have several `ShellCommand` objects with the same pid then destruct one of them, and imo better to do it with the fresh process. 

And handling some corner cases like: 
>Also it is possible that old command is present but already terminated and waited.

Will bring complicated logic for complicated cases. 

Also suppose that scenario is close to impossible, because  that's mean that [`vfork` returns](https://github.com/ClickHouse/ClickHouse/blob/33913bc4b2e18b64c36aaa794b652d28ac9b40ff/src/Common/ShellCommand.cpp#L163) duplicate of the pid (obvious  OS bug) or we made duplicate of the `unique_ptr`.

Reply:
If old command is present and has `wait_called=true` I think it is better to replace it.

Otherwise, because two commands have the same PID:
- We should not kill this PID at all while destructing one of them
Or
- Kill this PID and remove old entry from collection (to eliminate terminated)

Reply:
Also it is possible that SIGCHLD comes before we add command object to the holder map. And we will not wait it.

Reply:
Tried to fix the scenarios that you have described.